### PR TITLE
Add an RPC timeout configuration for baseRPCClient and include timeout settings for op-node

### PR DESCRIPTION
**Pull request by boyuan-chen**
**Description**

The timeout settings in the` baseRPCClient` are hardcoded to 10 seconds and to 20 seconds for batch calls. This causes issues when the `op-node` sends RPC requests to the node and the response takes longer than 10 seconds to return. The issue arises because `erigon`, serving as the sequencer node, sometimes takes more than 30 seconds to return the `eth_getProof` request.

This PR addresses several aspects:

1. It allows for modification of the `baseRPCClient` timeout settings away from the default values.
2. It enables the adjustment of the timeout settings for the RPC endpoint of the op-node.
3. The op-node includes settings to configure their respective timeouts.

**Tests**

The unit test is added.

### CODE REVIEW

File: op-node/flags/flags.go [hunk 1]
```diff
@@ -57,6 +69,30 @@ var (
 		EnvVars: prefixEnvVars("RPC_PORT"),
 		Value:   9545, // Note: op-service/rpc/cli.go uses 8545 as the default.
 	}
+	RPCListenReadTimeout = &cli.DurationFlag{
```
Comment (trianglesphere) [inline op-node/flags/flags.go:77]: What's the benefit to having so many configuration options here? Are these set by default

File: op-node/flags/flags.go [hunk 2]
```diff
@@ -34,6 +34,18 @@ var (
 		Usage:   "Address of L2 Engine JSON-RPC endpoints to use (engine and eth namespace required)",
 		EnvVars: prefixEnvVars("L2_ENGINE_RPC"),
 	}
+	L2RpcTimeout = &cli.DurationFlag{
```
Comment (trianglesphere) [inline op-node/flags/flags.go:38]: Is this L2 specific or just a RPC / Batch RPC timeout?

File: op-service/dial/dial.go [hunk 3]
```diff
@@ -34,8 +34,8 @@ func DialEthClientWithTimeout(ctx context.Context, timeout time.Duration, log lo
 }
 
 // DialRollupClientWithTimeout attempts to dial the RPC provider using the provided URL.
-// If the dial doesn't complete within timeout seconds, this method will return an error.
-func DialRollupClientWithTimeout(ctx context.Context, timeout time.Duration, log log.Logger, url string) (*sources.RollupClient, error) {
+// If the dial and rpc calls don't complete within timeout seconds, this method will return an error.
+func DialRollupClientWithTimeout(ctx context.Context, timeout time.Duration, log log.Logger, url string, rpcTimeoutCfg ...client.BaseRPCTimeout) (*sources.RollupClient, error) {
```
Comment (trianglesphere) [inline op-service/dial/dial.go:38]: This is a weird place to configure timeouts for `sources.RollupClient`. I see that this is where we are creating it, but if we need to configure it, we should pass a configuration option to it.

File: op-node/flags/flags.go [hunk 4]
```diff
@@ -34,6 +34,18 @@ var (
...
+		Name:    "l2.rpc-timeout",
+		Usage:   "Timeout for L2 RPC requests",
+		EnvVars: prefixEnvVars("L2_RPC_TIMEOUT"),
+		Value:   time.Second * 10,
+	}
+	L2RpcBatchTimeout = &cli.DurationFlag{
+		Name:    "l2.rpc-batch-timeout",
+		Usage:   "Timeout for L2 RPC batch requests",
```
Comment (tynes) [inline op-node/flags/flags.go]: Why does it need to be 2 different flags? A batch request is still a request right?

Comment (boyuan-chen) [reply op-node/flags/flags.go]: In the referenced section of the code, https://github.com/ethereum-optimism/optimism/blob/3b28485044c183e56a7cfe0b535eb8b247f7fbc7/op-service/client/rpc.go#L166-L176
the timeout configuration for batch RPC calls is set to differ from that of single RPC calls.

File: op-node/flags/flags.go (hunk 1 above)
Comment (boyuan-chen) [reply op-node/flags/flags.go:77]: Yes, these values are set to defaults. However, the default timeout may be insufficient for RPC requests that take a long time. For instance, if the `eth_getProof` request exceeds 30 seconds, the `op-node`'s response to `optimism_outputAtBlock` will result in an error.

File: op-node/flags/flags.go (hunk 2 above)
Comment (boyuan-chen) [reply op-node/flags/flags.go:38]: This option is passed to the `BaseRPCClient`, so I believe it acts as a timeout for both RPC and Batch RPC.

File: op-service/dial/dial.go (hunk 3 above)
Comment (boyuan-chen) [reply op-service/dial/dial.go:38]: The `sources.RollupClient` acts as a wrapper for the `baseRPCClient`, which contains the timeout settings. Where do you think is the appropriate place to add the configuration?

**Comment by trianglesphere** [comment]
What is the minimal set of timeouts that need to be configured @Boyuan-Chen? This seems like a useful feature, but I'd like to minimize the amount of configuration/flags.

**Comment by boyuan-chen** [comment]
> What is the minimal set of timeouts that need to be configured @Boyuan-Chen? This seems like a useful feature, but I'd like to minimize the amount of configuration/flags.

All configurations have default values, so there is no need to set any if you wish to use the defaults. The RPC response time for the op-node and the timeout for the BaseRPCClient structure are separate entities, each with its own distinct flags.

File: op-node/flags/flags.go (hunk 4 above)
Comment (tynes) [reply op-node/flags/flags.go]: I would be supportive of this if we used 1 flag for the timeouts here and passed the same value to both

Comment (boyuan-chen) [reply op-node/flags/flags.go]: Ok! I will use one value.

Comment (boyuan-chen) [reply op-node/flags/flags.go]: @tynes The rpcBatchTimeout is removed!

**Comment by trianglesphere** [comment]
If there is a specific minimal set of configuration which is important, please open a PR with that set. I don't think we want to significantly expand the set of configuration here.

**Pull request closed without merging**